### DISCUSSION TITLE
Connect all up

### DISCUSSION THREAD

**Discussion by kurt-k3po** [techie66/EngineDataLogger#20] [Ideas]
Have you thought about using the Innovate ISP2 protocol for connecting all the pieces together? 
Especially for EngineDataLogger-Engine Data and the DashBoard? 
On DashBoard side, someone could use existing software to analyze and display the logged data. On EngineData Side it would be more general to adapt/develop own sensor controller.

**Comment by techie66** [comment]
Long term plan is to use CAN2 protocol to connect sensor controllers. I would have to look again to see if ISP2 is suitable for the dashboard.

**Reply by techie66** [reply]
After taking a look at the specs for ISP2, I'm willing to try adapting EngineData over to see how it works. It will require some work on libISP2 in order to support, but that is mostly abstraction related. I see potential problems since ISP2 is limited to 13-bit values, so that pretty much rules out floating point numbers, but that is what is already being done with EngineData.  On the dashboard side, I don't see a good reason to get rid of flatbuffers. It provides a lot of flexibility, is already implemented and tested, and supports any type of data I might want. Although probably not a real limitation, ISP2 has a maximum 128 number of 16-bit "channels" of which only 13 bits are useful.

Just to clarify one point. The eventual move to CAN has little bearing one the data protocol. If CAN gets implemented, it can still use ISP2 as the protocol for data. CAN would also provide a nice way to "label" which sensor data is actually included in the packet through PIDs.

**Reply by kurt-k3po** [reply]
O.k. CAN sounds also good to me. Is there a chance to work on that in the near of time?
I have a Ignitech WB2 Lambda Controller at hand which has CAN connectivity. At the moment, I use this WB2 with an Arduino based AFR gauge I build years ago. MCP2515 - CAN controller is available too (with SPI on PI side). 
As you know, I am not a good programmer, but electronics is fine for me. In June I have some track time with a race bike where I plan to log some data. If there is a chance to get the lambda running with the Ignitech, I put all effort in assemble the pieces and doing all the tests.

**Reply by techie66** [reply]
I'm actually working on it right now. Trying to get hardware set up for CAN. If you've used a MCP2515, I could use advice. Once I have the hardware set up I can start writing the code to pull data from the CAN bus. I have some track time in May so I'm trying to get it ready by then, but I don't want to break what's already working.

**Reply by kurt-k3po** [reply]
I use a CAN Module like this one: https://www.ebay.de/itm/MCP2515-CAN-Bus-Modul-mit-TJA1050-Transceiver-5V-For-Arduino-Raspberry-Pi-Neu/184678179380
Tell me, how can I support you. Is it the SPI side to the Raspberry? I can check some tutorials on how to connect such Interfaces to RPI and get it to work for example.

**Reply by kurt-k3po** [reply]
Btw. I'm doing all the RPI tests on a PI Zero W as this is the target platform I will use later on the bike. Is this o.k?
Ah, I see, the GPIO pins of RPi are not 5V tolerant, So I orderd 3.3v 5v level shifter first. In the meantime I try with a voltage divider.

**Reply by techie66** [reply]
I've ordered a board that is 3v3 compatible to try out, and I'm testing on a pi zero w as well. I want to put the mcp2515 on my custom board to simplify hooking everything up. If I'm reading the datasheet right, I only need a crystal and capacitors along with a few resistors to make it run so I will get that prototype made up soon. 
I anticipate that socketCAN will make getting the packets into the DataLogger pretty easy so by next week I hopefully can start trying to implement at least logging for the oxygen sensor.

**Reply by kurt-k3po** [reply]
Can you provide a link to the board you ordered? I will then give it a try also. You design a custom PCB for the logger? Interesting. 
All the CAN HW designs I know of uses the MCP2515 in combination with CAN Transceiver such as TJA1050. The TJA1050 converts the RXCAN / TXCAN of MCP2515 to the CANH CANL differential signal levels.

Ok. I see, using a MCP2562 as transceiver instead of TJA1050 
would be smarter, as this chip use 3,3V for VIO (to connect direct to MCP2515) and VDD 5V to drive CANH / CANL lines.

**Reply by techie66** [reply]
The board- https://www.amazon.com/gp/aw/d/B07VMB1ZKH?ref=ppx_pt2_mob_b_prod_image

Custom board for front controls - https://github.com/techie66/EngineDataLogger-FrontControls/tree/development/hardware

**Reply by kurt-k3po** [reply]
I tested the waveshare RS485 CAN HAT board right now. This board work without problem with following configuration, as this board have a 12 MHz oscillator.

/boot/config.txt:

dtparam=spi=on
dtoverlay=mcp2515-can0,oscillator=12000000,interrupt=25
dtoverlay=sp1-1cs

**Comment by kurt-k3po** [comment]
I just wired a MCP2515 CAN Module to SPI0 on RPi using 3,3V for a test.
Edit /boot/config.txt and add the following:

dtparam=spi=on
dtoverlay=mcp2515-can0,oscillator=8000000,interrupt=25
dtoverlay=sp1-1cs

Reboot RPi and start can0 interface in loopback mode to test
> sudo ip link set can0 up type can bitrate 500000 loopback on

then start 'candump can0' in one terminal and send  with 'cansend can0 001#1122334455667788' from a 2nd terminal.

kurt@k3po-pi:~ $ candump can0
  can0  001   [8]  11 22 33 44 55 66 77 88
  can0  001   [8]  11 22 33 44 55 66 77 88
  can0  001   [8]  11 22 33 44 55 66 77 88
  can0  001   [8]  11 22 33 44 55 66 77 88

Even the loopback test works. See tomorrow if I can get some data over the true CAN wires. Not sure, because the tranceiver need 5V and the GPIO pins of RPi are not 5V tolerant. Therefore had to wait until other board arrive

**Reply by kurt-k3po** [reply]
approved! After modification of the MCP2515 CAN module to accept 3,3V and 5V I am able to read CAN messages.
I sucessfully connected my Ignitech WB2 Lambda controller to RPi.

root@k3po-pi:~# candump can0
  can0  322   [8]  10 27 10 27 C4 09 DC 05
  can0  322   [8]  10 27 10 27 C4 09 DC 05
  can0  322   [8]  10 27 10 27 C4 09 DC 05
  can0  322   [8]  10 27 10 27 C4 09 DC 05
  can0  322   [8]  10 27 10 27 C4 09 DC 05

That means in general, CAN setup work. As the module you ordered uses the same GPIO as my module, I expect no further problems to get your hardware set up.

Regards

Kurt

**Reply by techie66** [reply]
what mode was the WB2 in when you got those packets? 

> 5) CAN data flow
> The controller includes data output via CAN bus. Here you can enter three kinds of data formats.
>> 1) Ignitech - optional ID notification (decimal), adjustable speed. Notification structure: 0.1
>> BYTE - resulting AFR x 10 ; 2.3 BYTE - resulting λ x 100 ; 4.5 BYTE – voltage on
>> OUT1[mV] ; 6.7 BYTE – voltage on OUT2[mV]. Little endian notification.
>> 2) J1939 – message defined by J1939-71,v pgn 65193 as spn1119 , speed optional, number
>> ECU (WB2 controller number) optional.
>> 3) Proprietal – user-defined notification.

**Reply by kurt-k3po** [reply]
Mode is 1. Ignitech.
Here is a picture of the config of the Ignitech WB2 Lambda configuration.
CAN ID is selectable. Here it is 0x322 (802 decimal)

![grafik](https://user-images.githubusercontent.com/74114807/115109023-d051a880-9f73-11eb-9efb-2e0eb31cd7f6.png)

This works with the  waveshare RS485 CAN HAT board too
https://www.amazon.com/gp/aw/d/B07VMB1ZKH?ref=ppx_pt2_mob_b_prod_image

**Reply by techie66** [reply]
Can you set mode to 2/J1939 and post output of: candump can0 

If possible, can you do that from a running engine?

**Reply by kurt-k3po** [reply]
Here is another log. I did this not on a running engine, but I put the LSU4.9 i a tube and burned with a lighter on the open end. So it was not free to air. This gets a lambda between approximately 0.7 and 1.2. I also add a timestamp to the candump output..

This is the Ignitech WB2 config I used:
![grafik](https://user-images.githubusercontent.com/74114807/115142298-4bcd5b80-a041-11eb-8b4a-5736515b2e42.png)

The following logfile recorded on RPi with:  _candump -t z can0_

[lambda_can.log](https://github.com/techie66/EngineDataLogger/files/6330994/lambda_can.log)

Hope, this helps. If I should get the whole stuff to a real running engine, let me know. But this will take some time.

**Reply by techie66** [reply]
Thanks for the log file Kurt. I'm working on this right now. J1939 is more difficult than I anticipated. the Ignitech protocol is much simpler. J1939 does not give a lambda value, but % oxygen by volume. I am not sure how to convert from O2 % to lambda, so I will get the Ignitech format working first. If you can help me convert the O2 % to lambda, I can get J1939 support too.

**Reply by techie66** [reply]
I've pushed a new branch 'feature-can' with very basic support to get lambda from mode 1 "Ignitech" CAN frames. It isn't pretty, probably has bugs, but I'd like you to test it and see if it works. It will only work if you keep the CAN ID set to 802 (0x322)

**Reply by kurt-k3po** [reply]
Jacob, this is really great, thank you for that.
I just cloned the 'feature-can' branch and RPi is compiling at the moment. Make it sense, to install the bluetooth and flatbuffer at this time for this branch? What can be displayed at the 'dashboard' with this branch?
Also, I try to use the RPi GPIO 14+15 with RS232 converter for serial connection to TCIP4. I don't want to use the USB port for that.
My attempt is to get the lambda and TCIP4 logged to csv logfile and display some info on front control at runtime.

ok. Had a quick look to the code. Is it possible to get the 'advance' logged from TCIP4, too? If this could be easily implemented, I would really look forward to this. Then all parameters could be logged  which are needed to work on the 3D mapping of the Ignitech ignition in the first time.

**Reply by techie66** [reply]
Bluetooth and flatbuffers are necessary for the dashboard. Have a look at cb1100f-app.fbs to see what is displayed. The app is not very configurable at the moment so pretty much everything is always shown. It is fairly easy to log the advance from the tcip4, I will start adding that. Please let me know how testing goes with the CAN interface.

**Comment by ivanzanzani** [comment]
You two are too far ahead ... hats off !

**Comment by kurt-k3po** [comment]
Jacob, by the way - it's Ignitech **WB2** , not Ignitech WC-2 ;-)

**Reply by techie66** [reply]
Oops. 😆 I was tired.

**Comment by techie66** [comment]
This discussion has now started to center around CAN support. Let's continue with #4 existing issue.
